### ISSUE TITLE
IO tests fail with DomainMismatch on 32-bit architectures

### ISSUE THREAD

**Issue by ViviCoder** [GraphBLAS/python-suitesparse-graphblas#48]
When running tests on a 32-bit Debian distribution (i386, for instance), some IO tests fail with `suitesparse_graphblas.exceptions.DomainMismatch`.
See the full log below.
```
============================= test session starts ==============================
platform linux -- Python 3.10.6, pytest-7.1.2, pluggy-1.0.0+repack
rootdir: /<<PKGBUILDDIR>>, configfile: setup.cfg
collected 7 items

suitesparse_graphblas/tests/test_doctest.py .                            [ 14%]
suitesparse_graphblas/tests/test_exceptions.py .                         [ 28%]
suitesparse_graphblas/tests/test_io.py FF.                               [ 71%]
suitesparse_graphblas/tests/test_package.py .                            [ 85%]
suitesparse_graphblas/tests/test_scalar.py .                             [100%]

=================================== FAILURES ===================================
____________________________ test_serialize_matrix _____________________________

    def test_serialize_matrix():
        T = lib.GrB_INT64
        A = matrix.new(T, 2, 2)
        for args in zip(*_test_elements(T)):
            f = _element_setters[T]
            check_status(A, f(A[0], *args))
        data = matrix.serialize(A)
>       B = matrix.deserialize(data)

suitesparse_graphblas/tests/test_io.py:98: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
suitesparse_graphblas/io/serialize.py:94: in deserialize_matrix
    check_status(
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

obj = <cdata 'struct GB_Matrix_opaque *' NULL>, response_code = -5

    def check_status(obj, response_code):
        """Check the return code of the GraphBLAS function.
    
        If the operation was successful, return None.
    
        If the operation returned no value return `exceptions.NoValue`.
    
        Otherwise it is an error, lookup the exception and the error
        description, and throw the exception.
    
        """
        if response_code == GrB_SUCCESS:
            return
        if response_code == GrB_NO_VALUE:
            return ex.NoValue
    
        if ffi.typeof(obj).item.kind == "pointer":
            obj = obj[0]
    
        cname = ffi.typeof(obj).cname
        error_func = _error_func_lookup.get(cname)
        if error_func is None:
            raise TypeError(f"Unknown cname {cname} looking up error string.")
    
        string = ffi.new("char**")
        error_func(string, obj)
        text = ffi.string(string[0]).decode()
>       raise _error_code_lookup[response_code](text)
E       suitesparse_graphblas.exceptions.DomainMismatch

suitesparse_graphblas/__init__.py:175: DomainMismatch
____________________________ test_serialize_vector _____________________________

    def test_serialize_vector():
        T = lib.GrB_INT64
        v = vector.new(T, 3)
        check_status(v, lib.GrB_Vector_setElement_INT64(v[0], 2, 0))
        check_status(v, lib.GrB_Vector_setElement_INT64(v[0], 10, 1))
        data = vector.serialize(v, lib.GxB_COMPRESSION_LZ4HC, level=7)
>       w = vector.deserialize(data)

suitesparse_graphblas/tests/test_io.py:121: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
suitesparse_graphblas/io/serialize.py:122: in deserialize_vector
    check_status(
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

obj = <cdata 'struct GB_Vector_opaque *' NULL>, response_code = -5

    def check_status(obj, response_code):
        """Check the return code of the GraphBLAS function.
    
        If the operation was successful, return None.
    
        If the operation returned no value return `exceptions.NoValue`.
    
        Otherwise it is an error, lookup the exception and the error
        description, and throw the exception.
    
        """
        if response_code == GrB_SUCCESS:
            return
        if response_code == GrB_NO_VALUE:
            return ex.NoValue
    
        if ffi.typeof(obj).item.kind == "pointer":
            obj = obj[0]
    
        cname = ffi.typeof(obj).cname
        error_func = _error_func_lookup.get(cname)
        if error_func is None:
            raise TypeError(f"Unknown cname {cname} looking up error string.")
    
        string = ffi.new("char**")
        error_func(string, obj)
        text = ffi.string(string[0]).decode()
>       raise _error_code_lookup[response_code](text)
E       suitesparse_graphblas.exceptions.DomainMismatch

suitesparse_graphblas/__init__.py:175: DomainMismatch
=========================== short test summary info ============================
FAILED suitesparse_graphblas/tests/test_io.py::test_serialize_matrix - suites...
FAILED suitesparse_graphblas/tests/test_io.py::test_serialize_vector - suites...
========================= 2 failed, 5 passed in 0.98s ==========================
```

**Comment by eriknw** [comment]
Thanks for the report.  That's unexpected, but then I don't think we've tested much on 32-bit.  How did you install/build SuiteSparse:GraphBLAS, and what version?

It appears to be generating an error from one of three locations here:
https://github.com/DrTimothyAldenDavis/GraphBLAS/blob/93cafcb63f5600fff4f8301bf4d29e9564b6dbdf/Source/GB_deserialize.c#L99-L124

Since we are serializing/deserializing with INT64, a builtin type, and we pass NULL to deserialize as the datatype, I expect it's the first GrB_DOMAIN_MISMATCH.

@ViviCoder can you try installing `python-graphblas` (e.g., `pip install python-graphblas` or `conda install -c conda-forge python-graphblas`) and running its tests via `pytest --pyargs graphblas -k serialize`?

Other avenues to explore:
 - print the vales of `data` in the test (or in deserialize) so we can work through the deserialize logic "by hand"
 - try to do a simple serialize/deserialize in C to see if the problem is with `python-suitesparse-graphblas` or SuiteSparse:GraphBLAS

@DrTimothyAldenDavis do you support and/or test on 32-bit Debian distro, i386 arch?  (I know 32-bit Windows is unsupported).  Any ideas why this is failing?  The test is doing a straightforward serialize/deserialize of a matrix and vector with INT64 dtype, and passes NULL as the datatype to deserialize.

One of us may be able to try to reproduce the error and try to debug ourselves.  I won't be able to look into this until next week at the earliest.

**Comment by ViviCoder** [comment]
SuiteSparse;GraphBLAS 7.2.0 was built from sources on i386 arch, and the generated Debian package was installed.
However, it was not tested directly because the tests require MATLAB (or Octave) to run.
@DrTimothyAldenDavis By the way, it would be nice to have direct tests of the C API.

Here is the result of `pytest --pyargs graphblas -k serialize`:
```
============================= test session starts ==============================
platform linux -- Python 3.10.7, pytest-7.1.2, pluggy-1.0.0+repack
rootdir: /
collecting ... Running tests with "suitesparse" backend, blocking=True, record=False, mapnumpy=False, runslow=False
collected 656 items / 650 deselected / 6 selected                              

usr/local/lib/python3.10/dist-packages/graphblas/tests/test_matrix.py F  [ 16%]
usr/local/lib/python3.10/dist-packages/graphblas/tests/test_pickle.py ss [ 50%]
ss                                                                       [ 83%]
usr/local/lib/python3.10/dist-packages/graphblas/tests/test_vector.py F  [100%]

=================================== FAILURES ===================================
______________________________ test_ss_serialize _______________________________

A = "M_0"      nvals  nrows  ncols  dtype   format
gb.Matrix     12      7      7  INT64  bitmapr

    def test_ss_serialize(A):
        for compression, level, nthreads in itertools.product(
            [None, "none", "default", "lz4", "lz4hc"], [None, 1, 5, 9], [None, -1, 1, 10]
        ):
            if level is not None and compression != "lz4hc":
                with pytest.raises(TypeError, match="level argument"):
                    A.ss.serialize(compression, level, nthreads=nthreads)
                continue
            a = A.ss.serialize(compression, level, nthreads=nthreads)
>           C = Matrix.ss.deserialize(a, nthreads=nthreads)

usr/local/lib/python3.10/dist-packages/graphblas/tests/test_matrix.py:3576: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
usr/local/lib/python3.10/dist-packages/graphblas/_ss/matrix.py:4106: in deserialize
    check_status_carg(
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

response_code = -5, type_name = 'Matrix'
carg = <cdata 'struct GB_Matrix_opaque *' NULL>

    def check_status_carg(response_code, type_name, carg):
        if response_code == GrB_SUCCESS:
            return
        if response_code == GrB_NO_VALUE:  # pragma: no cover
            return NoValue
        try:
            error_func = libget(f"GrB_{type_name}_error")
        except AttributeError:  # pragma: no cover
            text = (
                f"Unable to get the error string for type {type_name}.  "
                "This is most likely a bug in graphblas.  Please report this as an issue at:\n"
                "    https://github.com/python-graphblas/python-graphblas/issues\n"
                "Thanks (and sorry)!"
            )
        else:
            string = ffi.new("char**")
            error_func(string, carg)
            text = ffi.string(string[0]).decode()
>       raise _error_code_lookup[response_code](text)
E       graphblas.exceptions.DomainMismatch

usr/local/lib/python3.10/dist-packages/graphblas/exceptions.py:136: DomainMismatch
______________________________ test_ss_serialize _______________________________

v = "v_0"      nvals  size  dtype  format
gb.Vector      4     7  INT64  bitmap

    def test_ss_serialize(v):
        for compression, level, nthreads in itertools.product(
            [None, "none", "default", "lz4", "lz4hc"], [None, 1, 5, 9], [None, -1, 1, 10]
        ):
            if level is not None and compression != "lz4hc":
                with pytest.raises(TypeError, match="level argument"):
                    v.ss.serialize(compression, level, nthreads=nthreads)
                continue
            a = v.ss.serialize(compression, level, nthreads=nthreads)
>           w = Vector.ss.deserialize(a, nthreads=nthreads)

usr/local/lib/python3.10/dist-packages/graphblas/tests/test_vector.py:2239: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
usr/local/lib/python3.10/dist-packages/graphblas/_ss/vector.py:1668: in deserialize
    check_status_carg(
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

response_code = -5, type_name = 'Vector'
carg = <cdata 'struct GB_Vector_opaque *' NULL>

    def check_status_carg(response_code, type_name, carg):
        if response_code == GrB_SUCCESS:
            return
        if response_code == GrB_NO_VALUE:  # pragma: no cover
            return NoValue
        try:
            error_func = libget(f"GrB_{type_name}_error")
        except AttributeError:  # pragma: no cover
            text = (
                f"Unable to get the error string for type {type_name}.  "
                "This is most likely a bug in graphblas.  Please report this as an issue at:\n"
                "    https://github.com/python-graphblas/python-graphblas/issues\n"
                "Thanks (and sorry)!"
            )
        else:
            string = ffi.new("char**")
            error_func(string, carg)
            text = ffi.string(string[0]).decode()
>       raise _error_code_lookup[response_code](text)
E       graphblas.exceptions.DomainMismatch

usr/local/lib/python3.10/dist-packages/graphblas/exceptions.py:136: DomainMismatch
```
Does it help?

**Comment by DrTimothyAldenDavis** [comment]
I haven’t tested it in i386.

A complete set of tests on a pure C API is impossible. I have to check the
functionality of GraphBLAS and doing so requires me to write it twice
because it’s a very complex API. There’s no norm (Ax-b) for example to
quickly check results. So I rely on MATLAB/Octave for the 2nd
implementation.

I could add more tests in pure C though. Like serialize/deserialize which
is easy to test in pure C.

On Sat, Sep 17, 2022 at 5:09 AM ViviCoder ***@***.***> wrote:

> SuiteSparse;GraphBLAS 7.2.0 was built from sources on i386 arch, and the
> generated Debian package was installed.
> However, it was not tested directly because the tests require MATLAB (or
> Octave) to run.
> @DrTimothyAldenDavis <https://github.com/DrTimothyAldenDavis> By the way,
> it would be nice to have direct tests of the C API.
>
>
> --
Sent from Gmail Mobile

**Comment by eriknw** [comment]
> Here is the result of pytest --pyargs graphblas -k serialize:

Thanks for sharing.  This give the same error going through a different code path in Python.  We're not doing anything especially weird in either of those code paths.

I think the next thing to do is to try this in C.  I'll try to share C code to test this by the end of this week if it's still needed (sorry for the delay, I'm stretched thin this week).

**Comment by eriknw** [comment]
I'm glad to hear this is also fixed in 7.4.0.0. Closing.
